toucan_connectors/facebook_ads/facebook_ads_connector.py

```diff
@@ -0,0 +1,141 @@
+            "and <a href='https://developers.facebook.com/docs/marketing-api/reference/ad-campaign-group/ads/'>for all the ads under a specific account"
+        ),
+    )
+
+    def determine_url(self) -> str:
+        format_key_mapping = {
+            FacebookadsDataKind.campaigns: {'act_id': self.parameters.get('account_id')},
+            FacebookadsDataKind.all_ads: {'act_id': self.parameters.get('account_id')},
+            FacebookadsDataKind.ads_under_campaign: {
+                'campaign_id': self.parameters.get('campaign_id')
+            },
+        }
+        return urljoin(
+            API_BASE_ROUTE,
+            API_ENDPOINTS_MAPPING[self.data_kind].format(**format_key_mapping[self.data_kind]),
+        )
+
+    def determine_query_params(self) -> Dict[str, str]:
+        params = {}
+
+        if self.data_fields:
+            params['fields'] = self.data_fields
+
+        for k, v in self.parameters.items():
+            params[k] = v
+
+        return params
+
+
+class FacebookAdsConnector(ToucanConnector):
+    _auth_flow = 'oauth2'
+    auth_flow_id: Optional[str]
+
+    data_source_model: FacebookadsDataSource
+
+    def __init__(self, **kwargs) -> None:
+        super().__init__(
+            **{k: v for k, v in kwargs.items() if k not in OAuth2Connector.init_params}
+        )
+        self.__dict__['_oauth2_connector'] = OAuth2Connector(
+            auth_flow_id=self.auth_flow_id,
+            authorization_url=AUTHORIZATION_URL,
+            scope=SCOPES,
+            token_url=TOKEN_URL,
+            secrets_keeper=kwargs['secrets_keeper'],
+            redirect_uri=kwargs['redirect_uri'],
+            config=OAuth2ConnectorConfig(
+                client_id=kwargs['client_id'],
+                client_secret=kwargs['client_secret'],
+            ),
+        )
+
+    @staticmethod
+    def get_connector_secrets_form() -> ConnectorSecretsForm:
+        return ConnectorSecretsForm(
+            documentation_md=(Path(os.path.dirname(__file__)) / 'doc.md').read_text(),
+            secrets_schema=OAuth2ConnectorConfig.schema(),
+        )
+
+    def retrieve_tokens(self, authorization_response: str):
+        """
+        In the Hubspot oAuth2 authentication process, client_id & client_secret
+        must be sent in the body of the request so we have to set them in
+        the mother class. This way they'll be added to her get_access_token method
+        """
+        return self.__dict__['_oauth2_connector'].retrieve_tokens(authorization_response)
+
+    def build_authorization_url(self, **kwargs):
+        return self.__dict__['_oauth2_connector'].build_authorization_url(**kwargs)
+
+    def _get_access_token(self):
+        return self.__dict__['_oauth2_connector'].get_access_token()
+
+    @staticmethod
+    def _handle_data_pagination(url: str, params: Dict) -> List[Dict]:
+        response = None
+        response_data = None
+        data = []
+
+        while not response or has_next_page(response_data):
```

Comment:
I there a way to fall in an endless loop if the network is unreachable ? 
Shouldn't the loop break after n retries ?

Reply:
The loop should break at the first raised error:
```
>>> requests.get("http://urlquinexistepas.fr")
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/urllib3/connection.py", line 159, in _new_conn
    conn = connection.create_connection(
  File "/usr/lib/python3/dist-packages/urllib3/util/connection.py", line 61, in create_connection
    for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
  File "/usr/lib/python3.8/socket.py", line 918, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -2] Name or service not known
```